src/Core/src/Hosting/ImageSources/ImageSourceToImageSourceServiceTypeMapping.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.Maui.Hosting
+{
+	internal sealed class ImageSourceToImageSourceServiceTypeMapping
+	{
+		private static readonly ConcurrentDictionary<IImageSourceServiceCollection, ImageSourceToImageSourceServiceTypeMapping> s_instances = new();
+
+		internal static ImageSourceToImageSourceServiceTypeMapping GetInstance(IImageSourceServiceCollection collection) =>
+			s_instances.GetOrAdd(collection, static _ => new ImageSourceToImageSourceServiceTypeMapping());
+
+		private readonly object _lock = new();
+		private readonly List<(Type ImageSource, Type ImageSourceService)> _typeMapping = new(8); // MAUI registers 8 image source services at startup
```

Comment:
This makes me wonder why we need to register both the concrete ImageSource classes and their interfaces? At least it seems like for `FileImageSource`, `StreamImageSource`, and `FontImageSource` we wouldn't need to register those if types can be resolved to the service registered with their interface.

However, for `UriImageSource : IUriImageSource, IStreamImageSource`, we would need to retain that default `AddService` call to ensure `MyCustomType : UriImageSource` properly grabs `UriImageSourceService` instead of `StreamImageSourceService`.

This would actually be an instance of having matches that don't derive from each other, when we previously had the "ambiguous" exception. I haven't looked through all of the tests yet, but we should try to cover such case.

Reply:
> This makes me wonder why we need to register both the concrete ImageSource classes and their interfaces? 

I don't think we necessarily need to register all of them. The `UriImageSource` is probably a case where it is necessary, otherwise we could get incorrect matches.

We have some tests that cover these "ambiguous" scenarios. Maybe there is a scenario we don't test yet?